src/com/sun/jna/Cleaner.java

```diff
@@ -0,0 +1,152 @@
+/* Copyright (c) 2021, Matthias Bläsing, All Rights Reserved
+ *
+ * The contents of this file is dual-licensed under 2
+ * alternative Open Source/Free licenses: LGPL 2.1 or later and
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ *
+ * You can freely decide which license you want to apply to
+ * the project.
+ *
+ * You may obtain a copy of the LGPL License at:
+ *
+ * http://www.gnu.org/licenses/licenses.html
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ *
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+
+package com.sun.jna;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Implement ReferenceQueue based cleanup of resources associated with GCed
+ * objects. It replaces the {@code Object#finalize} based resource deallocation
+ * that is deprecated for removal from the JDK.
+ *
+ * <p><strong>This is not intended for public use.</strong> On JDK 9+ use
+ * {@link java.lang.ref.Cleaner}.</p>
+ */
+public class Cleaner {
+    private static final Cleaner INSTANCE = new Cleaner();
+
+    static Cleaner getCleaner() {
+        return INSTANCE;
+    }
+
+    private final ReferenceQueue<Object> referenceQueue;
+    private final Thread cleanerThread;
+    private CleanerRef firstCleanable;
+
+    private Cleaner() {
+        referenceQueue = new ReferenceQueue<Object>();
+        cleanerThread = new Thread() {
+            @Override
+            public void run() {
+                while(true) {
+                    try {
+                        Reference<? extends Object> ref = referenceQueue.remove();
+                        if(ref instanceof CleanerRef) {
+                            ((CleanerRef) ref).clean();
+                        }
+                    } catch (InterruptedException ex) {
```

Comment:
Should we set `Thread.currentThread.interrupt()` here?  I recognize that the thread will essentially terminate on an interrupt and it probably doesn't matter in practice, but will keep static analysis tools happier.

Reply:
I don't get the intention. InterruptedException is raised if the waiting thread is interrupted (in this case the cleaner thread). Why would an analyser expect that the threads interrupts itself again?

Reply:
Analyzers don't like when you catch an interrupted exception and do nothing.  The general guidance is to either re-throw the exception or re-set the interrupted flag so that any upstream code recognizes the interruption.

As I said it probably doesn't matter in this case as the thread just terminates. (I'm not sure if it would terminate with a different exit code with that flag or not, or if that matters.)

This article says what I'm saying a lot better than I can:  https://programming.guide/java/handling-interrupted-exceptions.html